docs/design/resource-qos.md

```diff
@@ -0,0 +1,224 @@
+
+## Requests and Limits
+
+For each resource, containers can specify a resource request and limit, 0 <= request <= limit <= Infinity.
+If the container is successfully scheduled, the container is guaranteed the amount of resource requested.
+Scheduling is based on `requests` and not `limits`.
+The container will not be allowed to exceed the specified limit.
+How the request and limit are enforced depends on whether the resource is [compressible or incompressible](resources.md).
+
+### Compressible Resource Guarantees
+
+- For now, we are only supporting CPU.
+- Containers are guaranteed to get the amount of CPU they request, they may or may not get additional CPU time (depending on the other jobs running).
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 60% of the CPU, and container B requests for 30% of the CPU. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10% of CPU will be distributed to A and B in a 2:1 ratio (implementation discussed in later sections).
+- Containers will be throttled if they exceed their limit. If limit is unspecified, then the containers can use excess CPU when available.
+
+### Incompressible Resource Guarantees
+
+- For now, we are only supporting memory.
+- Containers will get the amount of memory they request, if they exceed their memory request, they could be killed (if some other container needs memory), but if containers consume fewer resources than requested, they will not be killed (except in cases where system tasks or daemons need more memory).
+- Containers will be killed if they use more memory than their limit.
+
+### Kubelet Admission Policy
+
+- Pods will be admitted by Kubelet based on the sum of requests of its containers. The Kubelet will ensure that sum of requests of all containers (over all pods) is within the system’s resource (for both memory and CPU).
+
+## QoS Classes
+
+In an overcommitted system (where sum of requests > machine capacity) containers might eventually have to be killed, for example if the system runs out of CPU or memory resources. Ideally, we should kill containers that are less important. For each resource, we divide containers into 3 QoS classes: *Guaranteed*, *Burstable*, and *Best-Effort*, in decreasing order of priority.
+
+The relationship between "Requests and Limits" and "QoS Classes" is subtle. Theoretically, the policy of classifying containers into QoS classes is orthogonal to the requests and limits specified for the container. Hypothetically, users could use an (currently unplanned) API to specify whether a container is guaranteed or best-effort. However, in the current design, the policy of classifying containers into QoS classes is intimately tied to "Requests and Limits" - in fact, QoS classes are used to implement some of the memory guarantees described in the previous section.
+
+Containers can be of one of 3 different classes:
```

Comment:
I think we should look to update kubectl describe pod to capture this change.

Right now, we describe the qos tier on a per resource basis (which I think we should keep), but we do not roll that up on a container wide tier.